File: docs/snippets/all/views/spatial3d.py
"""Use a blueprint to customize a Spatial3DView."""

import rerun as rr
import rerun.blueprint as rrb
from numpy.random import default_rng

rr.init("rerun_example_spatial_3d", spawn=True)

# Create some random points.
rng = default_rng(12345)
positions = rng.uniform(-5, 5, size=[50, 3])
colors = rng.uniform(0, 255, size=[50, 3])
radii = rng.uniform(0.1, 0.5, size=[50])

rr.log("points", rr.Points3D(positions, colors=colors, radii=radii))
rr.log("box", rr.Boxes3D(half_sizes=[5, 5, 5], colors=0))

# Create a Spatial3D view to display the points.
blueprint = rrb.Blueprint(
    rrb.Spatial3DView(
        origin="/",
        name="3D Scene",
        # Set the background color to light blue.
        background=[100, 149, 237],
        # Configure the eye controls.
        eye_controls=rrb.EyeControls3D(
            position=(0.0, 0.0, 2.0),
            look_target=(0.0, 2.0, 0.0),
            eye_up=(-1.0, 0.0, 0.0),
            spin_speed=0.2,
            kind=rrb.Eye3DKind.FirstPerson,
            speed=20.0,
        ),
        # Configure the line grid.
        line_grid=rrb.LineGrid3D(
            visible=True,  # The grid is enabled by default, but you can hide it with this property.
            spacing=0.1,  # Makes the grid more fine-grained.
            # By default, the plane is inferred from view coordinates setup, but you can set arbitrary planes.
            plane=rr.components.Plane3D.XY.with_distance(-5.0),
            stroke_width=2.0,  # Makes the grid lines twice as thick as usual.
            color=[255, 255, 255, 128],  # Colors the grid a half-transparent white.
        ),
        spatial_information=rrb.SpatialInformation(
            target_frame="tf#/",
            show_axes=True,
            show_bounding_box=True,
        ),
    ),
    collapse_panels=True,
)

rr.send_blueprint(blueprint)

File: docs/snippets/all/views/spatial3d.rs
//! Use a blueprint to customize a Spatial3DView.

use rerun::blueprint::{
    Blueprint, Spatial3DView, archetypes as blueprint_archetypes,
    components as blueprint_components,
};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let blueprint = Blueprint::new(
        Spatial3DView::new("3D Scene")
            .with_origin("/")
            .with_background(rerun::Color::from_rgb(100, 149, 237))
            .with_eye_controls(
                blueprint_archetypes::EyeControls3D::new()
                    .with_position((0.0, 0.0, 2.0))
                    .with_look_target((0.0, 2.0, 0.0))
                    .with_eye_up((-1.0, 0.0, 0.0))
                    .with_spin_speed(0.2)
                    .with_kind(blueprint_components::Eye3DKind::FirstPerson)
                    .with_speed(20.0),
            )
            .with_line_grid(
                blueprint_archetypes::LineGrid3D::new()
                    .with_visible(true)
                    .with_spacing(0.1)
                    .with_plane(rerun::components::Plane3D::new(
                        [0.0, 0.0, 1.0],
                        -5.0,
                    ))
                    .with_stroke_width(2.0)
                    .with_color([255, 255, 255, 128]),
            )
            .with_spatial_information(
                blueprint_archetypes::SpatialInformation::new("tf#/")
                    .with_show_axes(true)
                    .with_show_bounding_box(true),
            ),
    );

    let rec = rerun::RecordingStreamBuilder::new("rerun_example_spatial_3d")
        .with_blueprint(blueprint)
        .spawn()?;

    let positions = (0..50).map(|i| {
        let i = i as f64;
        [
            (i * 0.37).sin() * 4.0,
            (i * 0.21).cos() * 4.0,
            -5.0 + i * 10.0 / 49.0,
        ]
    });
    let colors = (0..50_i32).map(|i| {
        [
            ((i * 53) % 255) as u8,
            ((128 + i * 29) % 255) as u8,
            (255 - i * 17).rem_euclid(255) as u8,
        ]
    });
    let radii = (0..50).map(|i| (0.1 + (0.5 - 0.1) * i as f64 / 49.0) as f32);

    rec.log(
        "points",
        &rerun::Points3D::new(positions)
            .with_colors(colors)
            .with_radii(radii),
    )?;
    rec.log(
        "box",
        &rerun::Boxes3D::from_half_sizes([(5.0, 5.0, 5.0)]).with_colors([0]),
    )?;

    Ok(())
}
